docs/rfcs/env-vars.md

```diff
@@ -0,0 +1,216 @@
+familiar with YAML, because of the implicit type casting rules and the
+way we store intermediate values, we can get unintuitive results.
+
+The most clear example of this is issue
+[*#8565*](https://github.com/open-telemetry/opentelemetry-collector/issues/8565):
+When setting a variable to value `0123` and using it in a string-typed
+field, it will end up as the string `"83"` (where as the user would
+expect the string to be `0123`).
+
+#### We are less restrictive than the Configuration WG
+
+The Configuration WG defines an [*environment variable expansion feature
+for SDK
+configurations*](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#environment-variable-substitution).
+This accepts only non empty alphanumeric + underscore identifiers
+starting with alphabetic or underscore. If the Configuration WG were to
+expand this in the future (e.g. to include other features present in
+Bash-like syntax), we would not be able to expand our braces syntax to
+support new features without breaking users.
+
+## Desired behavior
+
+*This section is written as if the changes were already implemented.*
+
+The Collector supports **two** different syntaxes for environment
+variable resolution:
+
+1.  The *braces syntax*, `${ENV}`.
+2.  The *env provider syntax*, `${env:ENV}`.
+
+These both have **the same character set and behavior**. They both use
+the env provider under the hood.
+
+The naked syntax supported in BASH is not supported in the Collector.
+Escaping is supported by using two dollar signs. Escaping is also
+honored for unsupported identifiers like `${1}`.
+
+### Type casting rules
+
+The environment variable value is parsed by the yaml.v3 parser to an
+any-typed variable and the original representation as a string is stored
+for numeric types. The `yaml.v3` parser mostly follows the YAML v1.2
+specification with [*some
+exceptions*](https://github.com/go-yaml/yaml#compatibility). You can see
+how it works for some edge cases in this example:
+[*https://go.dev/play/p/RtPmH8aZA1X*](https://go.dev/play/p/RtPmH8aZA1X).
+
+When unmarshalling, we use mapstructure with WeaklyTypedInput
+**disabled**. We check via a hook an `AsString` method from confmap.Conf
+and use its return value when it is valid and we are mapping to a string
+field. This method has default casting rules for unambiguous scalar
+types but may return the original representation depending on the
+construction of confmap.Conf
+
+For using this notation in inline mode (e.g.`http://endpoint/${env:PATH}`), we
+use the `AsString` method from confmap.Conf.
+
+### Character set
+
+An environment variable identifier must be a nonempty ASCII alphanumeric
+or underscore starting with an alphabetic or underscore character. Its
+maximum length is 200 characters. Both syntaxes support recursive
+resolution.
+
+When an invalid identifier is found, a warning log is emitted, and the
+string is kept as-is.
```

Comment:
I am undecided about this, I feel like we should fail in some cases but I am not sure about which. open-telemetry/opentelemetry-specification/issues/3981 has some more discussion about this

Reply:
@TylerHelmuth The one edge case here is `${1}` from open-telemetry/opentelemetry-collector-contrib/issues/9984. Should we fail in that case and force uses to use escaping syntax?

Reply:
Since that is what we currently do and it seems impossible to detect all kinds of syntaxes past or future that would overlap with this, I have changed this to an error on 3f0546741415843f5dca369b80d263a239fb31f2

Reply:
@mx-psi I feel like `${1}` corresponds with the regex and should be expanded. If users need the literal `${1}` can they do `$${1}`?

Reply:
> I feel like since ${1} corresponds with the regex it should be expanded

I think we are on the same page, but to be clear `1` is not a valid identifier per the 'desired behavior' section (valid identifiers need to start with underscore or ASCII alphabetic characters), so it should **not** be expanded, but rather raise an error.

Reply:
Ah ok, then we are on the same page, I just got the outcome wrong. My opinion is the `${1}` should fail and if users need that literal string they should escape: `$${1}`.

This prevents us from having any edge cases with our unwrap syntax while keeping all functionality available.

Reply:
@mx-psi where in the spec do we need to raise this issue?

Reply:
Tyler and I discussed this offline but for the benefit of anybody else keeping track of this: this is being discussed on open-telemetry/opentelemetry-specification/pull/4002